### PR TITLE
OBPIH-7597 Allow mass Recipient Edit for line items in multiple POs

### PR DESCRIPTION
**Pull request by SebastianLib**
### :sparkles: Description of Change

[//]: <> (A concise summary of what is being changed. Please provide enough context for reviewers to be able to understand the change and why it is necessary.)

**Link to GitHub issue or Jira ticket:**
https://pihemr.atlassian.net/browse/OBPIH-7597

**Description:**

---
### :camera: Screenshots & Recordings (optional)

[//]: <> (If this PR contains a UI change, consider attaching one or more screenshots or recordings to help reviewers visualize the change. Otherwise, you can remove this section.)

### CODE REVIEW

File: grails-app/services/org/pih/warehouse/data/PersonService.groovy [hunk 1]
```diff
@@ -146,4 +146,33 @@ class PersonService {
...
+     */
+    Person getActivePerson(String recipient) {
+        recipient = recipient?.trim()
+        if (StringUtils.isBlank(recipient)) {
+            return null
+        }
+
+        try {
+            InternetAddress internetAddress = new InternetAddress(recipient, false)
+            internetAddress.validate()
```
Comment (alannadolny) [inline grails-app/services/org/pih/warehouse/data/PersonService.groovy]: I don't understand that part

File: grails-app/services/org/pih/warehouse/importer/PurchaseOrderActualReadyDateImportDataService.groovy [hunk 2]
```diff
@@ -101,7 +102,12 @@ class PurchaseOrderActualReadyDateImportDataService implements ImportDataService
                     command.errors.reject("Row ${index + 1}: Could not parse date ${params['actualReadyDate']} on 'Actual Ready Date'. Expected date format: ${Constants.EXPIRATION_DATE_FORMAT}")
                 }
             }
-
+            if (!StringUtils.isBlank(params["recipient"])) {
+                Person recipient = personService.getActivePerson(params["recipient"])
+                if (!recipient) {
+                    command.errors.reject("Row ${index + 1}: Recipient ${params["recipient"]} does not exist in OpenBoxes. Please check your data")
```
Comment (alannadolny) [inline grails-app/services/org/pih/warehouse/importer/PurchaseOrderActualReadyDateImportDataService.groovy]: you can assign the params["recipient"] to a variable

File: grails-app/services/org/pih/warehouse/data/PersonService.groovy (hunk 1 above)
Comment (SebastianLib) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy]: Here we use InternetAddress only to detect whether the value entered in the recipient column is an email address (e.g. `alannadolny@gmail.com` or `Alan Nadolny <alannadolny@gmail.com>`). Based on that, we decide whether to check the person by email (getActivePersonByEmail method) or by full name (getActivePersonByName method). You can read more about this here: https://docs.oracle.com/javaee/5/api/javax/mail/internet/InternetAddress.html

I was inspired by @ewaterman  code:
```
    Person getOrCreatePersonByRecipient(String recipient) {
        if (StringUtils.isBlank(recipient)) {
            return null
        }

        InternetAddress internetAddress
        try {
            internetAddress = new InternetAddress(recipient, false)
        }
        catch (AddressException ignored) {
            // If recipient isn't a valid internet address, it must be a regular name.
            return getOrCreatePersonFromNames(recipient)
        }

        // If a person exists with the given email, return them. Note that this can return inactive people so
        // the caller must properly handle that case. We do this to avoid creating duplicate users.
        Person person = getPersonByEmail(internetAddress.address)
        if (person) {
            return person
        }

        // Otherwise create a new person.
        if (!internetAddress.personal) {
            throw new RuntimeException("Cannot save new recipient without a name: ${recipient}")
        }
        String[] names = extractNames(internetAddress.personal)
        person = new Person(firstName: names[0], lastName: names[1], email: internetAddress.address)
        if (!person.save(flush: true)) {
            throw new ValidationException("Cannot save recipient ${recipient} due to errors", person.errors)
        }
        return person
    }
```

Comment (ewaterman) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy]: I forgot I wrote that 😅 I think the code in PersonService can be refactored so that we have cleaner methods:
- getPersonByName(String) -> split into firstname, lastname and query
- getPersonByEmail(String) -> simply query on Person.email
- getPerson(String) -> calls both getPersonByEmail and getPersonByName to find the Person
- getActivePerson(String) -> calls getPerson then returns them if active == true
- getOrCreatePerson(String) -> calls getPerson and if null, creates a Person

but I won't make you do that.

Comment (SebastianLib) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy]: The refactor looks simple enough, so I can try tomorrow

Comment (alannadolny) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy]: btw, the internet address looks weird to me, and I am confused every time I see that code :smile:  Maybe we can try to do some kind of validator for that? (https://grails.apache.org/docs/7.0.4/ref/Constraints/email.html)

@ewaterman

Comment (ewaterman) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy]: the problem is that when we import from a csv or xls file, we have just one field representing the Person (in this case "recipient"). And to be user friendly, we want to accept either name or email in that field. So we can't just wrap the field in a simple built-in validator. The String field that we get is either an email or a name and we need to figure out which it is here

File: grails-app/services/org/pih/warehouse/data/PersonService.groovy [hunk 3]
```diff
@@ -146,4 +122,43 @@ class PersonService {
...
+    /**
+     * Find an active person by recipient string which can be either an email address or a full name.
+     *
+     * @param recipient Either a RFC822 internet/email address, a plain email address, or a full name.
+     * @return An active Person matching the recipient, or null if not found or person is inactive.
+     */
+    Person getActivePerson(String recipient) {
+        Person person = getPerson(recipient)
+        return person?.active ? person : null
+    }
```
Comment (ewaterman) [inline grails-app/services/org/pih/warehouse/data/PersonService.groovy:162]: looks good! I know you didn't introduce it, but "recipient" doesn't make much sense here IMO because this can be used by any flow, not just orders and shipments. Really it's just a string for identifying a person.

Can you please change the "recipient" variables here to "identifier" or "personIdentifier" or something similar?

File: grails-app/services/org/pih/warehouse/data/PersonService.groovy [hunk 4]
```diff
@@ -146,4 +122,43 @@ class PersonService {
...
+            internetAddress.validate()
+            return getPersonByEmail(internetAddress.address)
+        } catch (AddressException ignored) {
+            // If recipient isn't a valid internet address, it must be a regular name.
+            // Check if name has at least two words
+            if (recipient.split(" ").length < 2) {
+                return null
+            }
+            return getPersonByName(recipient)
+        }
```
Comment (ewaterman) [inline grails-app/services/org/pih/warehouse/data/PersonService.groovy:149]: super nitpick: Because you return in your try block, you can leave the catch block empty and put this code outside of it. The code doesn't actually relate to handling the thrown exception so IMO it's cleaner to have it outside the block. Up to you though

File: src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy [hunk 5]
```diff
@@ -101,4 +62,37 @@ class PersonServiceSpec extends Specification implements ServiceUnitTest<PersonS
...
+
+        expect:
+        service.getActivePerson(recipient)?.id == expectedId
+
+        where:
+        recipient                              || expectedId
+        'Justin Miranda'                       || '1'
+        'Unknown Name <justin@openboxes.com>'  || '1'
+        'Justin Inactive'                      || null
+    }
```
Comment (ewaterman) [inline src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:108]: Nice tests! I feel you could be more exhaustive with inputs here. Both these methods could test:
```
- 'Justin'
- 'Justin Miranda'
- 'Justin Justy Miranda'
- 'justin@openboxes.com'
- 'unknown@openboxes.com'
- 'Justin Miranda <justin@openboxes.com>'
- 'Justin Miranda <unknown@openboxes.com>'
- 'Unknown Name <justin@openboxes.com>'
- 'Justin Inactive'
- null
```

File: grails-app/services/org/pih/warehouse/data/PersonService.groovy (hunk 3 above)
Comment (SebastianLib) [reply grails-app/services/org/pih/warehouse/data/PersonService.groovy:162]: yes, I think that "identifier" makes more sense. I will change that

File: src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy (hunk 5 above)
Comment (SebastianLib) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:108]: I agree it’s a good idea to test more cases. I haven’t written many tests in my career, so thanks for the guidance on what exactly to test :rocket:

Comment (alannadolny) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:108]: 👏

File: src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy [hunk 6]
```diff
@@ -101,4 +62,48 @@ class PersonServiceSpec extends Specification implements ServiceUnitTest<PersonS
...
+
+        where:
+        recipient                                || expectedId
+        'Justin'                                 || null
+        'Justin Miranda'                         || '1'
+        'Justin Justy Miranda'                   || null
+        'justin@openboxes.com'                   || '1'
+        'unknown@openboxes.com'                  || null
+        'Justin Miranda <justin@openboxes.com>'  || '1'
+        'Justin Miranda <unknown@openboxes.com>' || null
```
Comment (SebastianLib) [inline src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:82]: I think that this test case is curious. In this test case: `Justin Miranda <unknown@openboxes.com>` the result is null. This happens because the method detects an email format and only searches by the email address. Even though the name 'Justin Miranda' is correct, the system doesn't find the person because the email is unknown. 

On the other hand, if the email is correct but the name is wrong (e.g. `Unknown Name <justin@openboxes.com>`), the system finds the person. 

Maybe somehow we should improve this? If the email search fails, then we should also check by name or maybe we should stay as it is and always prioritize email :thinking:

Comment (ewaterman) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:82]: hmm it's an interesting point. I don't know if there's a clear answer.

Another argument is that if you provide both, both should match. Person isn't actually unique on email or on name so just because the name or email match, it doesn't mean you're getting the user that you want. Providing both could mean you want to be very specific. But idk, that could be overly restrictive.

I think it's more important that the email match because email far more likely to be unique. Two people can have the same name, but I'd be surprised if we have two people with the same email (unless the same person is in the system multiple times).

So given that, I guess the current logic makes sense to prioritize email, but I suppose we could have it try name afterwards if the email doesn't match.

That said, I don't actually know if anyone is providing strings like `Justin Miranda <unknown@openboxes.com>`. I imagine 99% of the time it's either just name or just email. So maybe it's not a problem that we need to deal with

Comment (SebastianLib) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:82]: @kchelstowski @alannadolny what do you think about this?

Comment (alannadolny) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:82]: I believe it's not a problem because it's a hardcore edge case. But maybe as a simple solution, we can show the user an error like: "We found a person by the name that you provided, but the email doesn't match. Please check your data."

**Review by alannadolny — APPROVED** [review]

File: src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy (hunk 6 above)
Comment (SebastianLib) [reply src/test/groovy/org/pih/warehouse/person/PersonServiceSpec.groovy:82]: hmmmm I think it's better to leave it as it is :thinking:

**Review by kchelstowski — APPROVED** [review]
